### ISSUE TITLE
API to activate first parent dir with Project.toml

### ISSUE THREAD

**Issue by fonsp** [JuliaLang/Pkg.jl#2640]
When launching `julia` with argument `--project`, then it will _"search through parent directories until a Project.toml or JuliaProject.toml file is found."_ ([docs](https://docs.julialang.org/en/v1/manual/command-line-options/)). This is really useful for projects with scripts is nested folders. 

It would be great if this was available as a **function** in Pkg, to be called at the top of a script/notebook. This would be useful in environments where setting the `project` flag is impractical (Pluto.jl, Jupyter, VS Code). It also helps to simplify the reproducibility steps, since the standard `julia` command works.

I would be happy to make this PR myself if there is interest?

**Comment by Datseris** [comment]
To give some more context, the proposed functionality is already implemented in DrWatson, as a function `quickactivate` and a macro `@quickactivate`. https://juliadynamics.github.io/DrWatson.jl/dev/project/#Activating-a-Project-1  

You start a script with the lines:
```julia
using DrWatson
@quickactivate "ProjectName" 
```
and Julia starts scanning the folder that contains your script until it finds a project, and activates that project. The argument `"ProjectName"` only checks that the activated project matches the given name, and gives an error if not.

The Macro is the most useful version, and our users are extremely happy with it. It allows for reproducible scientific projects, allows finding your "correct" project automatically by running your script (no manual labor!), and it allows for easily switching projects within the same Julia session.

The only problem it has is that it requires `DrWatson` to be loaded. This means that `DrWatson` needs to be available on the global environment to use `@quickactivate`. That's a problem because even if the Project.toml of your scientific project contains `DrWatson`, this doesn't really guarantee that the global environment of the person you've sent it to will also have it. 

This as a result also leads to some difficulties in Pluto.jl notebooks, cf https://github.com/JuliaDynamics/DrWatson.jl/issues/261 . Given that this functionality seems generally useful, we think it is a good idea to have it in Pkg.jl directly.

**Comment by fredrikekre** [comment]
You can use `Pkg.activate(Base.current_project())`

**Comment by Datseris** [comment]
Yes (and actually the source of this function was the starting point for `quickactivate`). But if we really care about reproducibility, then `Base.current_project()` would have to be documented (currently it has no docstring) and be formally made part of the public API. If not, it could "potentially" change or be removed at any point between even patch releases (according to SemVer stuff). 

Does this sound like something possible for you? If yes, then this issue practically resolves to a documentation PR.

**Comment by aplavin** [comment]
> This would be useful in environments where setting the project flag is impractical (Pluto.jl, Jupyter, VS Code).

How is it relevant to Jupyter and VS Code? Jupyter (IJulia) already activates the closest parent env:

> The default Jupyter kernel that is installed by IJulia starts with the Julia command line flag --project=@.. A Project.toml (or JuliaProject.toml) in the folder of a notebook (or in a parent folder of this notebook) will therefore automatically become the active project for that notebook.

VS Code suggests to choose which of the parent envs to activate when opening a project for the first time, and activates it automatically afterwards.

**Comment by fonsp** [comment]
Hello! Maybe we could decide whether to add this API or close this issue? The suggested API is:

```julia
Pkg.activate(search_parents=true)
```

To have behaviour identical to `julia --project`: *"search through parent directories until a Project.toml file is found."*

**Comment by Datseris** [comment]
+1. Unfortunately, making this the default would be a breaking change :(

**Comment by Datseris** [comment]
Actually, it may be breaking no matter what, because currently if one calls `Pkg.activate` without input arguments then it says:

> If no argument is given to activate, then use the first project found in LOAD_PATH

**Comment by Datseris** [comment]
Additionally, `Pkg.activate(search_parents=true)` would somehow need to know that it should search the parents of the directory it was called in.

Doesn't just porting `@quickactivate` to Pkg satisfy all our needs? Would this be an option for the Pkg developers?

**Comment by fredrikekre** [comment]
Perhaps `Pkg.activate("@.")` since that is already used for this in `JULIA_PROJECT`?

**Comment by Datseris** [comment]
Seems to me like a simple enough solution!

**Comment by KristofferC** [comment]
I don't really like magical strings like that in normal julia functions. Fine for the REPL mode.

**Comment by Datseris** [comment]
> [@Datseris](https://github.com/Datseris) I think searching relative to `pwd()` makes sense, just like `Pkg.activate("..")` is relative to `pwd()`, not to its call location.

I disagree; I would argue what makes most sense is the lines that contains the statement, i.e., `@__DIR__`. In most scenarios where you care about managing dedicated Project.toml per project (such as in a scientific context), your scripts are in a folder that also contains your project in a parent directory. Whether this folder coincides with `pwd()` is secondary. I think it's fair to say that it is much more common that your script is in your project directory than it is that your `pwd()` matches the project you want to activate. 

As an example, I have 8 projects in my VSCode "workspace" and when VSCode opens it sets my `pwd()` to the first project. 7/8 times that's not the directory I would like to have for searching where my project file is.

**Comment by Datseris** [comment]
I've suggested porting `@quickactivate` here before and faced unsupported decline. Can someone **clearly, and objectively**, tell me why porting the `@quickactivate` macro here isn't the solution to every single concern raised so far? "I don't like it" is not an objective argument.
